include/ygm/random.hpp

```diff
@@ -0,0 +1,75 @@
+// Copyright 2019-2021 Lawrence Livermore National Security, LLC and other YGM
+// Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <ygm/comm.hpp>
+
+#include <random>
+
+namespace ygm {
+namespace random {
+
+/// @brief Applies a simple offset to the specified seed according to rank index
+/// @tparam ResultType The random number (seed) type
+/// @param comm The ygm::comm to be used
+/// @param seed The specified seed
+/// @return simply returns seed + rank
+template <typename ResultType>
+ResultType simple_offset(ygm::comm &comm, ResultType seed) {
+  return seed + comm.rank();
+}
+
+/// @brief Applies no change to the specified seed
+/// @tparam ResultType The random number (seed) type
+/// @param comm The ygm::comm to be used
+/// @param seed The specified seed
+/// @return simply returns the unmodified seed
+template <typename ResultType>
+ResultType no_offset(ygm::comm &comm, ResultType seed) {
+  return seed;
+}
+
+/// @brief A wrapper around a per-rank random engine that manipulates each
+///        rank's seed according to a specified strategy
+/// @tparam RandomEngine The underlying random engine, e.g. std::mt19337
+/// @tparam Function A function on (ygm::comm, result_type) -> result_type that
+///         modifies seeds for each rank
+template <typename RandomEngine,
+          typename RandomEngine::result_type (*Function)(
+              ygm::comm &, typename RandomEngine::result_type)>
+class random_number_generator {
```

Comment:
I think we should use the term engine here.   ygm::random_engine to mirror std::default_random_engine